Description: Patch out intersphinx
 Avoids internet access during build
Author: Thomas Goirand <zigo@debian.org>
Date: Fri, 4 Aug 2017 21:47:21 +0200
Bug-Debian: https://bugs.debian.org/840092
Forwarded: no
Last-Update: 2024-02-14

diff --git a/doc/source/conf.py b/doc/source/conf.py
index cda080f..5b7889d 100644
--- a/doc/source/conf.py
+++ b/doc/source/conf.py
@@ -25,7 +25,6 @@
     'sphinx.ext.autodoc',
     'sphinx.ext.todo',
     'sphinx.ext.coverage', 
-    'sphinx.ext.intersphinx',
     'sphinxcontrib.apidoc'
 ]
 
