Description: Install all files
Author: Thomas Goirand <zigo@debian.org>
Forwarded: no
Last-Update: 2024-09-20

--- python-eventlet-0.36.1.orig/MANIFEST.in
+++ python-eventlet-0.36.1/MANIFEST.in
@@ -2,3 +2,4 @@ recursive-include tests *.py *.crt *.key
 recursive-include doc *.rst *.txt *.py Makefile *.png
 recursive-include examples *.py *.html
 include CONTRIBUTING.md MANIFEST.in NEWS AUTHORS LICENSE README.rst SECURITY.md tox.ini
+recursive-include eventlet *
